### PR TITLE
Self Service: Create App

### PR DESCRIPTION
**Pull request by wira-mms**

### CODE REVIEW

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs [hunk 1]
```diff
@@ -152,6 +153,41 @@ public async Task AddCompanyAppSubscriptionAsync(Guid appId, string userId)
...
+                Name = appInputModel.Title,
+                MarketingUrl = appInputModel.ProviderUri,
+                AppUrl = appInputModel.AppUri,
+                ThumbnailUrl = appInputModel.LeadPictureUri,
+                ContactEmail = appInputModel.ContactEmail,
+                ContactNumber = appInputModel.ContactNumber,
+                ProviderCompanyId = appInputModel.ProviderCompanyId,
+                AppStatusId = PortalBackend.PortalEntities.Enums.AppStatusId.CREATED
+            };
+            await this.context.Apps.AddAsync(appEntity).ConfigureAwait(false);
```
Comment (ntruchsess) [inline src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs:173]: According to the documentation AddAsync should only be used if value generators are being used that require it:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.addasync?view=efcore-6.0
In all other cases the use of synchronous Add is recommended.
(same with 'AddRangeAsync' which is being used few lines below)

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs [hunk 2]
```diff
@@ -152,6 +153,41 @@ public async Task AddCompanyAppSubscriptionAsync(Guid appId, string userId)
...
+            await this.context.AppLicenses.AddAsync(appLicenseEntity);
+            await this.context.SaveChangesAsync();                
+
+            await this.context.AppAssignedLicenses.AddAsync(new AppAssignedLicense(appEntity.Id, appLicenseEntity.Id));
+            await this.context.AppAssignedUseCases.AddRangeAsync(appInputModel.UseCaseIds.Select(uc => new AppAssignedUseCase(appEntity.Id, uc)));
+            await this.context.AppDescriptions.AddRangeAsync(appInputModel.Descriptions.Select(d => new AppDescription(appEntity.Id, d.LanguageCode, d.LongDescription, d.ShortDescription)));
+            await this.context.AppLanguages.AddRangeAsync(appInputModel.SupportedLanguageCodes.Select(c => new AppLanguage(appEntity.Id, c)));
+            await this.context.SaveChangesAsync();
+
+            await transaction.CommitAsync();
```
Comment (ntruchsess) [inline src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs:186]: Why doing several calls to the database within an explicitly controlled transaction and not just build all related entities in memory and call save only once (which does the transaction-management implicitly)?

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs [hunk 3]
```diff
@@ -152,6 +153,41 @@ public async Task AddCompanyAppSubscriptionAsync(Guid appId, string userId)
...
+                ThumbnailUrl = appInputModel.LeadPictureUri,
+                ContactEmail = appInputModel.ContactEmail,
+                ContactNumber = appInputModel.ContactNumber,
+                ProviderCompanyId = appInputModel.ProviderCompanyId,
+                AppStatusId = PortalBackend.PortalEntities.Enums.AppStatusId.CREATED
+            };
+            await this.context.Apps.AddAsync(appEntity).ConfigureAwait(false);
+            await this.context.SaveChangesAsync();
+
+            var appLicenseEntity = new AppLicense(Guid.Empty, appInputModel.Price);
```
Comment (ntruchsess) [inline src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs:176]: Guid.Empty?

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs [hunk 4]
```diff
@@ -152,6 +153,41 @@ public async Task AddCompanyAppSubscriptionAsync(Guid appId, string userId)
             await this.context.SaveChangesAsync().ConfigureAwait(false);
         }
 
+        /// <inheritdoc/>
+        public async Task<Guid> CreateAppAsync(AppInputModel appInputModel)
+        {
+            using var transaction = this.context.Database.BeginTransaction();
+
+            // Add app to db
+            var appEntity = new PortalBackend.PortalEntities.Entities.App(Guid.Empty, appInputModel.Provider, DateTimeOffset.UtcNow)
```
Comment (ntruchsess) [inline src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs:162]: Why Guid.Empty?
There is no guid-generation set up in the database (yet). If you want to introduce such we should discuss this in a separate call.
Guids can be generated by the program using Guid.NewGuid()

File: src/marketplace/CatenaX.NetworkServices.App.Service/Controllers/AppsController.cs [hunk 5]
```diff
@@ -69,6 +70,30 @@ public async Task<ActionResult<AppDetailsViewModel>> GetAppDetailsByIdAsync([Fro
...
+        /// <returns>ID of created application.</returns>
+        /// <remarks>Example: POST: /api/apps</remarks>
+        /// <response code="201">Returns created app's ID.</response>
+        [HttpPost]
+        [Route("")]
+        [Authorize(Roles = "add_app")]
+        [ProducesResponseType(typeof(Guid), StatusCodes.Status201Created)]
+        public async Task<ActionResult<Guid>> CreateAppAsync([FromBody] AppInputModel appInputModel)
+        {
+            try
```
Comment (ntruchsess) [inline src/marketplace/CatenaX.NetworkServices.App.Service/Controllers/AppsController.cs:86]: please use the GeneralHttpErrorHandler (project CatenaX.NetworkServices.Framework.ErrorHandling) and remove the logging / statuscode-mapping boilerplate code (as it is done e.g. here: https://github.com/catenax-ng/product-portal-backend/blob/30f9db58e66ec40cb73a007e73a5b09a690fadf9/src/registration/CatenaX.NetworkServices.Registration.Service/Controllers/RegistrationController.cs#L121 )
To enable just add the GeneralHttpErrorHandler to the middleware in program.cs:
https://github.com/catenax-ng/product-portal-backend/blob/30f9db58e66ec40cb73a007e73a5b09a690fadf9/src/registration/CatenaX.NetworkServices.Registration.Service/Program.cs#L128

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/CatenaX.NetworkServices.App.Service/Controllers/AppsController.cs (hunk 5 above)
Comment (wira-mms) [reply src/marketplace/CatenaX.NetworkServices.App.Service/Controllers/AppsController.cs]: https://github.com/catenax-ng/product-portal-backend/pull/37/commits/ad9459cc8fe1adec5c1dbfd5a5cf952d0db39f39

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs (hunk 4 above)
Comment (wira-mms) [reply src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs]: https://github.com/catenax-ng/product-portal-backend/pull/37/commits/2826aca1596684a3c35e02f075af784d9cf6964e

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs (hunk 3 above)
Comment (wira-mms) [reply src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs]: https://github.com/catenax-ng/product-portal-backend/pull/37/commits/2826aca1596684a3c35e02f075af784d9cf6964e

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs (hunk 2 above)
Comment (wira-mms) [reply src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs]: https://github.com/catenax-ng/product-portal-backend/pull/37/commits/2826aca1596684a3c35e02f075af784d9cf6964e

File: src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs (hunk 1 above)
Comment (wira-mms) [reply src/marketplace/CatenaX.NetworkServices.App.Service/BusinessLogic/AppsBusinessLogic.cs]: https://github.com/catenax-ng/product-portal-backend/pull/37/commits/2826aca1596684a3c35e02f075af784d9cf6964e

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**